### PR TITLE
Distinguish ReplicationIT from a Publish end to end test

### PR DESCRIPTION
**Pull request by amit-jain**
## Description

- Create a standalone test to test replication with only status available on author.
- Rename ReplicationIT to PublishEndToEndIT - This test is intended to test the overall flow for publishing a page, the page being accessible on publish.

## Related Issue

https://github.com/adobe/aem-test-samples/issues/32

## Motivation and Context

The current ReplicationIT is an actual end to end test but the name suggests that the responsibility of the failure is for 1 component only. But that is not true as the assertions on presence/absence of a page for ADD/DELETE case relies on the page being accessible through the dispatcher and publish.

## How Has This Been Tested?

Changes tested on a cloud environment

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

**Comment by amit-jain** [comment]
@dulvac @cschneider 
Appreciate your review on this.

**Comment by cschneider** [comment]
@amit-jain Something is wrong in this PR. The ReplicationIT extends PublishIT but the other class is named PublishEndtoEndIT.

**Comment by amit-jain** [comment]
@cschneider 
> Something is wrong in this PR. The ReplicationIT extends PublishIT but the other class is named PublishEndtoEndIT.

Right, somehow that change was left uncommitted. Fixed now.

**Review by cschneider — APPROVED** [review]
LGTM

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
Comment (dulvac) [inline smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java:95]: Thanks for catching this. I'll open an issue in sling

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
Comment (dulvac) [inline smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java]: rename this to `waitPublishQueueEmptyOfPath`

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/ReplicationIT.java
Comment (dulvac) [inline smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/ReplicationIT.java]: I don't think we should do this. No base classes for tests as things get confusing really quickly. Whenever we have shared code, we put this in `Rule`s or utility classes.

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
Comment (dulvac) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java:95]: But why not make the methods like `activateAndDelete` throw the AssumtionViolatedException?

**Review by dulvac — CHANGES_REQUESTED** [review]
@amit-jain, mostly blocking because we don't want to have any base classes in these tests + minor changes suggested

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java]: Sorry couldn't notice the difference

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/ReplicationIT.java
Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/ReplicationIT.java]: Hmm...I think we will still have lot of shared code e.g the test methods and the internal methods they delegate to. Is that  acceptable? Here it ensures there is no duplication

Comment (dulvac) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/ReplicationIT.java]: Yes, it's ok to have shared code, but not through inheritance :)

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
Comment (dulvac) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java]: sry! :D Corrected the name above

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java:95]: This was existing code so, I did not catch this :)
But I still don't understand your point. The point is that any exception would be wrapped as a TimeoutException in the `Polling.poll()` method if invoked directly.

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/ReplicationIT.java
Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/ReplicationIT.java]: Sure. what I mean is that we will have duplicated code e.g. the methods below can't logically be placed anywhere common other than a test but I guess that's ok.
- testActivateAndDelete
- activateAndDelete

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
Comment (dulvac) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java:95]: I'm saying you can add this try/ catch block in methods like `activateAndDelete` and the rest, instead of the test method. But this is a minor thing

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/ReplicationIT.java
Comment (dulvac) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/ReplicationIT.java]: You can put them in a `Rule` class or a utility class, you don't need to duplicate the code

**Comment by amit-jain** [comment]
@dulvac Thanks for sticking by. I have updated the PR with the inheritance removal.

**Comment by amit-jain** [comment]
@dulvac Appreciate your review. Have updated the PR to create a rule instead as it makes additions for https://github.com/adobe/aem-test-samples/pull/39 easier and also removes lot of duplicated code still present before

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
Comment (dulvac) [inline smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java]: Don't call it `PublishRule` please. Maybe something like `ContentPublishRule`. It clashes with the base rules. e.g. `cqBaseClassRule.publishRule` and it's confusing

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/PublishRule.java
Comment (dulvac) [inline smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/PublishRule.java]: What is this for. Looking at the code, `isEmpty` is always `false` here. Remove?

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/PublishRule.java
Comment (dulvac) [inline smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/PublishRule.java]: 💯

**Review by dulvac — CHANGES_REQUESTED** [review]
Just a couple of more things, mainly the naming of the rule.
Thanks for the patience!

**Comment by dulvac** [comment]
Also @amit-jain , make sure to coordinate with https://github.com/adobe/aem-testing-clients/pull/59 (there is duplicate code between the two)

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java]: Sure..will change it

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/PublishRule.java
Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/PublishRule.java]: No, the value is retrieved from the json returned for each items in queues

**Comment by amit-jain** [comment]
> make sure to coordinate with adobe/aem-testing-clients#59 (there is duplicate code between the two)

Yes, that is required for and will be resolved with the https://github.com/adobe/aem-test-samples/pull/39

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/PublishRule.java
Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/PublishRule.java]: This is used to check if the upstream queue has items or not and if items available log if there's any error

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java]: Done!

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/ContentPublishRule.java
Comment (dulvac) [inline smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/ContentPublishRule.java:196]: @amit-jain , if `isEmpty` is `false`, it goes inside the `if` block

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/ContentPublishRule.java
Comment (dulvac) [inline smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/ContentPublishRule.java]: `isEmpty` is logged here, but it is always `false`, otherwise it does not go in the `if` block

Comment (dulvac) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/ContentPublishRule.java]: maybe you want this log line outside the `if(!isEmpty) {` block

Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/ContentPublishRule.java]: Aah! sorry it took time for me to understand that you were pointing to the log. Yes, makes sense only outside before even checking the value(for debugging)

Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/ContentPublishRule.java]: Done!

**Comment by amit-jain** [comment]
@tmaret Could you also please review

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
Comment (tmaret) [inline smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java:76]: Is the intent to cleanup resources in case of failed test ? If so, I think we should skip re-activating the page and simply cleanup with the `CQClient`.

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
Comment (tmaret) [inline smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java:103]: Same as above.

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/ContentPublishRule.java
Comment (tmaret) [inline smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/ContentPublishRule.java]: I think that the message should be more specific. We may report the actual status code received, specify that we are talking about the publish tier and suggest routing and access issues. I suggest the following

```
Failed to check the page {} via the AEM publish ingress (expected status {}, actual status {}). Please ensure that the CDN and Dispatcher configurations allow fetching the page.
```

Comment (tmaret) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/ContentPublishRule.java]: @dulvac  does the check go through the CDN ? If not, we'd obviously only mention the Dispatcher.

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/ContentPublishRule.java
Comment (tmaret) [inline smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/ContentPublishRule.java]: It looks like `skipDispatcherCache` is always `true`. It'd make sense to remove support for that parameter for simplicity sake.

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
Comment (tmaret) [inline smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java]: It'd make sense to clarify what we mean by end to end. A test page is published from author and publication is validated through the AEM publish ingress route (CDN > Dispatcher > AEM publish tier).

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/ReplicationIT.java
Comment (tmaret) [inline smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/ReplicationIT.java]: We may specify here that the test isolate the replication framework. A test page is published from author and publication is validated through the replication queues.

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
Comment (tmaret) [inline smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java:56]: We may align the comment to what the test really verifies

After activation, that the page is accessible via the publish ingress

After deactivation, that the page is no longer accessible via the publish ingress and that the queue is empty

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/ReplicationIT.java
Comment (tmaret) [inline smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/ReplicationIT.java]: This test has two parts, 1. activation and 2. delete. The first part is already tested via `testActivateAndDeactivate` and the second part is not part of the replication framework. So I think we should remove that test. The test suite will run faster without compromising coverage.

**Comment by tmaret** [comment]
Both `PublishEndToEndIT` and `ReplicationIT` should assert that the replication queue is empty before activating the page and throw `AssumptionViolatedException` if the queue is blocked.

**Comment by amit-jain** [comment]
@tmaret 

> Both PublishEndToEndIT and ReplicationIT should assert that the replication queue is empty before activating the page and throw AssumptionViolatedException if the queue is blocked.

Added a check which tries to assert empty queue or throw an exception in the before() method of the rule

**Comment by dulvac** [comment]
> Both PublishEndToEndIT and ReplicationIT should assert that the replication queue is empty before activating the page and throw AssumptionViolatedException if the queue is blocked.

@tmaret , @amit-jain please let's not do that. I won't weigh in into `ReplicationIT`, but for `PublishEndToEndIT` we want to make sure it tests publishing in an end-to-end fashion. If the replication queue is blocked let's say by a customer content package that doesn't install or for any other reason, we want the test to fail. These are not unit tests or integration tests (despite the confusing name :) ) and we rely on most of them to make sure business flows work and for the test to fail if they don't, no matter what the cause (of course, more specialized tests can be added, but we also need the e2e ones)

**Comment by tmaret** [comment]
> If the replication queue is blocked let's say by a customer content package that doesn't install or for any other reason, we want the test to fail

@dulvac @amit-jain We can make the assertion fail the test for the e2e one. We should keep the assertion though, because it clearly identifies the root cause and propose the solution. The error message should surface as much contextual information as possible, I suggest the following format

````
Can't run the test because the replication queue <queueName> is blocked by the package <packageId>. Please unblock the queue before running the test again.
````

**Review by tmaret — CHANGES_REQUESTED** [review]
Left a few comments inline.

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java:76]: This was an existent check in the tests. The intention is to retry till timeout is reached.

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java:103]: This was an existent check in the tests. The intention is to retry till timeout is reached.

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/ContentPublishRule.java
Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/ContentPublishRule.java]: Ack. Pre existing log message will change accordingly.

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/ContentPublishRule.java
Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/ContentPublishRule.java]: Ack.

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/ReplicationIT.java
Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/ReplicationIT.java]: .. replication queues on author.

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java:56]: ack

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/ReplicationIT.java
Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/ReplicationIT.java:84]: You suggest removing this test?

Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/ReplicationIT.java]: ok.

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
Comment (dulvac) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java:56]: @tmaret I think the point is to test that a delete in AEM triggers a deactivation of the page on publish. Which arguably is part of how replication is set up on the AEM deployment. So imo it would make sense to keep it.

**Comment by dulvac** [comment]
> > If the replication queue is blocked let's say by a customer content package that doesn't install or for any other reason, we want the test to fail
> 
> @dulvac @amit-jain We can make the assertion fail the test for the e2e one. We should keep the assertion though, because it clearly identifies the root cause and propose the solution. The error message should surface as much contextual information as possible, I suggest the following format
> 
> ```
> Can't run the test because the replication queue <queueName> is blocked by the package <packageId>. Please unblock the queue before running the test again.
> ```

That makes a lot of sense and I agree with 100% as long as the test still fails (with the advantage of a clear reason)

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java:56]: Yes, it makes sense to keep it for E2E test (didn't notice the comment was on that test). But remove it for ReplicationIT

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/ContentPublishRule.java
Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/rules/ContentPublishRule.java]: Updated, the logs after change will be as below. Since, the call `Poll` only returns a boolean we don't have access to the returned status. Will need a change in the polling method to add a custom exception and/or message.

> [main] INFO com.adobe.cq.cloud.testing.it.smoke.rules.ContentPublishRule - Checking page https://host1/content/test-site/testpage_6b132787-1d52-49b1-8af8-8a96948faf62.html returns status [404]
[main] WARN com.adobe.cq.cloud.testing.it.smoke.rules.ContentPublishRule - Failed to check the page /content/test-site/testpage_6b132787-1d52-49b1-8af8-8a96948faf62 via the AEM publish ingress (expected status [404]). Please ensure that the CDN and Dispatcher configurations allow fetching the page.

**Comment by amit-jain** [comment]
After the change, the logs and failure are logged as follows when the queue is blocked before the tests

> [main] WARN com.adobe.cq.cloud.testing.it.smoke.rules.ContentPublishRule - The replication queue c1439bf3-e622-4c3d-baff-6b8d031a3066-publishSubscriber is blocked by the item package-0@57591642.
......
[main] WARN com.adobe.cq.cloud.testing.it.smoke.rules.ContentPublishRule - The replication queue c1439bf3-e622-4c3d-baff-6b8d031a3066-publishSubscriber is blocked by the item package-0@57591642.
[main] WARN com.adobe.cq.cloud.testing.it.smoke.rules.ContentPublishRule - The replication queue c1439bf3-e622-4c3d-baff-6b8d031a3066-publishSubscriber is blocked by the item package-0@57591642.
[main] ERROR com.adobe.cq.cloud.testing.it.smoke.rules.ContentPublishRule - Failed to run test due to blocked queue. Please unblock the queue before running the test again
java.util.concurrent.TimeoutException: Call failed to return true in 60000 ms. Last exception was: null
	at org.apache.sling.testing.clients.util.poller.Polling.poll(Polling.java:133)
	at com.adobe.cq.cloud.testing.it.smoke.rules.ContentPublishRule.waitPublishQueueEmptyOfPath(ContentPublishRule.java:131)
	at com.adobe.cq.cloud.testing.it.smoke.rules.ContentPublishRule.before(ContentPublishRule.java:94)
  ......
[main] INFO com.adobe.cq.testing.junit.rules.Page - Deleted page at /content/test-site/testpage_31b90d8c-47bc-4c6a-875c-f2ca30fd1c1f
[ERROR] Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 66.567 s <<< FAILURE! - in com.adobe.cq.cloud.testing.it.smoke.ReplicationIT
[ERROR] testActivateAndDeactivate(com.adobe.cq.cloud.testing.it.smoke.ReplicationIT)  Time elapsed: 66.566 s  <<< ERROR!
java.util.concurrent.TimeoutException: Call failed to return true in 60000 ms. Last exception was: null

**Review by tmaret — APPROVED** [review]
LGTM, thanks @amit-jain !

File: smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java
Comment (dulvac) [inline smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java:73]: @amit-jain , I would do it the other way around and change the rule to throw an exception, please.

Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java:73]: @dulvac You want the meat of the method to move to the rule? What is the benefit you see for that? This seems to be some ad-hoc implementation to tide over a lacuna in the Poller right?

Comment (dulvac) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java:73]: @amit-jain , No, I'm talking about catching `AssumptionViolatedException` here from the `contentPublishRule`. Who throws the `AssumptionViolatedException` and why can't it throw an exception instead?

Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java:73]: From what I understand, the implementation wants to throw the `AssumptionViolatedException` to skip tests rather than fail in cases where the `AssumptionViolatedException` is thrown. The documentation over it also says that

> This is a workaround for correctly detecting assumption violations.

 The exception is thrown by the checkPage() method
`
       if (null != res && res.getStatusLine().getStatusCode() == SC_UNAUTHORIZED) {
            throw new AssumptionViolatedException("Publish requires auth for (SAML?). Skipping...");
        }
`

If we really don't care about this special handling then we can easily remove this and directly invoke the Poller.poll() warpping the method. wdyt?

Comment (dulvac) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java:73]: @amit-jain , we're conflating some use-cases here. We need the `AssumptionViolatedException` in checkPage to be there **and to propagate all the way through** (i.e. the test needs to be skipped in that case)

I was under the incorrect impression (my bad) that the rule throws an `AssumptionViolationException` if waitPublishQueueEmptyOfPath timed out. Which I see it's not the case and it should remain like that. However, we need to NOT catch this `AssumptionViolationException` thrown by `checkPage` in the test.

So to recap:
* if `checkPage` throws `AssumptionViolationException`, let it go through and **skip** the test
* if `waitPublishQueueEmptyOfPath` fails, **fail** the test with the clear message (which you've done iiuc)

Comment (amit-jain) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java:73]: @dulvac 
> However, we need to NOT catch this AssumptionViolationException thrown by checkPage in the test.

Yes that is also already the case. What might be confusing is the catch and rethrow. 
That as explained in the comments is to get over the limitation while wrapping the call in `Poller.poll()`. When `Poller.poll()` is callled it will after requisite retries throw only `TimeoutException` and not the underlying `AssumptionViolationException`. 

To get over that what the code is doing is making the 

- 1st call made as it is and rethrowing if its `AssumptionViolationException` to make it skip the test
- Retrying by wrapping the call in `Poller.poll()` when some other exception encountered. The call will eventually throw  a `TimeoutException` and fail if that other exception condition has not cleared.

Comment (dulvac) [reply smoke/src/main/java/com/adobe/cq/cloud/testing/it/smoke/PublishEndToEndIT.java:73]: @amit-jain , you're right, it seems to do what it should. Indeed, I did not see the fact that **the same** exception is thrown in case of `AssumptionViolationException`. Sorry about that!

**Comment by amit-jain** [comment]
@dulvac Thanks for reviewing. I believe the PR is ready to be merged as all review changes have been done.

**Comment by amit-jain** [comment]
@dulvac @tmaret gentle reminder for this PR!! What's the process to get this merged?

**Review by dulvac — APPROVED** [review]

**Comment by dulvac** [comment]
@amit-jain , I merged it now, thanks for the patience!

**Comment by amit-jain** [comment]
@dulvac Thank you so much!!

**Pull request merged**
